### PR TITLE
DataUnions: encapsulate contract handling, remove obsolete wrappers, improve validation, memberStats format

### PR DESCRIPTION
**Pull request by teogeb**
Split `DataUnionEndpoints.ts`: move contract handling to `Contracts.ts,` ABI definitions to `abi.ts` and inline public methods to `DataUnion.ts`

Smaller changes:
- Remove obsolete caching from address fetching
- Improve address validation (and add tests)
- Split tests, improve test cleanup
- Bug fix for payForSignatureTransport option in withdraws: method option overrides client option
- Format of `getMemberStats()` data modified (enum values changed, number values string->`BigNumber`) 
- Type annotations for `StreamrClientOptions`: many of the fields are required, as we apply a default value -> users give options as `Partial<StreamrClientOptions>`
- Id and debug are client properties, not options

### CODE REVIEW

**Review by jtakalai — CHANGES_REQUESTED** [review]
stellar job

File: test/integration/dataunion/calculate.test.ts
Comment (teogeb) [reply test/integration/dataunion/calculate.test.ts:38]: These lines have been moved into `afterAll()` block (lines 21-24) 
```
providerMainnet.removeAllListeners()
providerSidechain.removeAllListeners()
```

Comment (timoxley) [reply test/integration/dataunion/calculate.test.ts:38]: Yeah we should get rid of the `--forceExit` as it catches tests that didn't clean themselves up properly. If the test process doesn't exit then there's some async stuff leaking.

**Review by timoxley — COMMENTED** [review]
👍 structure better. lgtm. Get @jtakalai approval before merging.

File: src/Config.ts
Comment (teogeb) [reply src/Config.ts]: Fixed

File: src/dataunion/DataUnion.ts
Comment (teogeb) [reply src/dataunion/DataUnion.ts:541]: Fixed

File: src/Config.ts
Comment (teogeb) [reply src/Config.ts]: Removed `id` and `debug` fields from options as those were duplicate to `StreamrClient`'s fields. Neither of the fields affected the `id` and `debug` fields of the `StreamrClient` instance, so in practice those were not configuration options.

Changed `Connection` and `WebSocket` to get the debug instance from `StreamrClient`. 

`AuthFetch` creates its own debug instance. If we later refactor it to a class, it can also get the debug instance from `StreamrClient`.

File: src/dataunion/DataUnion.ts
Comment (teogeb) [reply src/dataunion/DataUnion.ts:516]: I've usually put the parameters in priority order to improve readability. If the most important details are at the start of the line, the reader can easily spot what is happening at the line. The start of the line needs to be read anyway, but there is no need to read the line fully if the reader is just browsing the code to gather some high-level understanding.

If minor details (or highly predictable parameter values) are put the ends of line lines, the reader can focus to those later, if needed. 

E.g. if someone reads only the bold parts in here, he/she should have quite a good understanding what the method does:

```
**If (!foobar) {**
    **const client = createClient()**
    **doSomething(foo, bar,** client)
    **somethingElse(foo, '123',** options, client)
} else {
    throw new Error()
}
```

Consistency is important, and the ability to curry functions. But here I'd like to trade those for the improved readability and "browseability". 

If there are a group of methods that use "pseudo-this" parameters, we could refactor those to a class, like in `Contracts.ts` in this PR.

File: src/dataunion/Contracts.ts
Comment (teogeb) [reply src/dataunion/Contracts.ts:79]: @jtakalai suggested earlier that DataUnion could initialize the contract objects when it is created (e.g. `getDataUnion()`). In that case DataUnion would contain the actual Contract objects as fields, not `contractAddress` and `sidechainAddress`. Then we could fetch `SidechainSigner` in `getDataUnion` and provide the signer to `getSidechainContract()` as a parameter.

File: src/dataunion/DataUnion.ts
Comment (jtakalai) [reply src/dataunion/DataUnion.ts:516]: I'd say +1 to having the "self" as first argument. It could be a separate refactor though.

File: test/integration/dataunion/calculate.test.ts
Comment (jtakalai) [reply test/integration/dataunion/calculate.test.ts:38]: for sure, that's the goal

**Review by jtakalai — APPROVED** [review]
looking good! Please add here (or into the ticket) a list of links to tickets of refactorings that were not done at this time but could be done later.

**Comment by teogeb** [comment]
Ticket for possible refactorings: https://linear.app/streamr/issue/NET-209/js-client-refactoring-of-dataunion-internals

**Pull request merged**
